### DISCUSSION TITLE
Clarification regarding override checking

### DISCUSSION THREAD

**Discussion by insilications** [microsoft/pyright#6265] [General]
I'm updating PySide6, which offers the official Python bindings for Qt, to offer primary support for pyright and a more complete python (strict) typing experience. I need clarification regarding how overriding is checked. I can't find any specific information in the PEP (https://peps.python.org/pep-0698/). The pyright docs states:

> reportIncompatibleMethodOverride [boolean or string, optional]: Generate or suppress diagnostics for methods that override a method of the same name in a base class in an incompatible manner (wrong number of parameters, incompatible parameter types, or incompatible return type). The default value for this setting is "none".

It states that wrong number of parameters, incompatible parameter types, or incompatible return type are all detected as incompatible overrides. Pylance detects the following as incompatible:

```
Method "event" overrides class "QAbstractAnimation" in an incompatible manner
  Parameter 2 name mismatch: base parameter is named "event", override parameter is named "e" (PylancereportIncompatibleMethodOverride)
```

```
class QPauseAnimation(PySide6.QtCore.QAbstractAnimation):
    @override
    def event(self, e: PySide6.QtCore.QEvent) -> bool: ...

class QAbstractAnimation(PySide6.QtCore.QObject):
    @override
    def event(self, event: PySide6.QtCore.QEvent) -> bool: ...

class QObject(Shiboken.Object):
    def event(self, event: PySide6.QtCore.QEvent) -> bool: ...
```

For parameters that can be used with keyword arguments, variable name checking is essential. But is it for positional-only parameters? Isn't it possible, in theory, for type checkers to differentiate between the two situations?
The **event** method in **QPauseAnimation** has the **QEvent** variable named as **e**. That is a direct consequence of the binding generator tool (Shiboken6), which reads the C++ headers and generates a function signature that is then used to generate stub files (.pyi). I can fix this in the stub files generation stage by telling it to rename to **event** or the same name as in the base class.

**Comment by insilications** [comment]
According to https://typing.readthedocs.io/en/latest/source/stubs.html#id22:

> Function and method definition syntax follows general Python syntax. Unless an argument name is prefixed with two underscores (but not suffixed with two underscores), it can be used as a keyword argument [2](https://typing.readthedocs.io/en/latest/source/stubs.html?highlight=enum#pep484):
> 
> ```
> # x is positional-only
> # y can be used positionally or as keyword argument
> # z is keyword-only
> def foo(__x, y, *, z): ...
> ```
> PEP 570 [7](https://typing.readthedocs.io/en/latest/source/stubs.html?highlight=enum#pep570) style positional-only parameters are currently not supported.


It seems that changing the **PySide6.QtCore.QEvent** argument to be prefixed by two underscores all across the inheritance hierarchy (PySide6.QtCore.QPauseAnimation -> PySide6.QtCore.QAbstractAnimation -> PySide6.QtCore.QObject) stops the PylancereportIncompatibleMethodOverride error. Can I use this to fix the error without any side effects?

**Reply by erictraut** [reply]
Python 3.8 and [PEP 570](https://peps.python.org/pep-0570/) introduced an official syntax to specify positional-only parameters using the `/` parameter. If your code doesn't need to run on Python interpreters prior to 3.8, then I recommend using the PEP 570 syntax.

For code bases that must continue to work with older versions of Python, the double underscore naming convention is supported by type checkers and linters as a backward-compatible way to indicate positional-only parameters. This is the technique used in typeshed stubs, for example.

**Reply by insilications** [reply]
Thanks. Can I apply the same fix, without side effects, to this case?

```
class QAbstractItemModel(QObject):
    def hasChildren(self, parent: PySide6.QtCore.QPersistentModelIndex = ..., /) -> bool: ...

class QAbstractTableModel(QAbstractItemModel):
    @override
    def hasChildren(self, parent: PySide6.QtCore.QPersistentModelIndex, /) -> bool: ...
```

Without marking **parent** as positional-only, I get:

> Method "hasChildren" overrides class "QAbstractItemModel" in an incompatible manner
>   Parameter 2 mismatch: base parameter has default argument value, override parameter does not Pylance(reportIncompatibleMethodOverride)


With positional-only, error goes away.
